src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,301 @@
+                    metrics = new Metrics
+                    {
+                        TotalTimeMS = (long)_intervalStopwatch.GetElapsedTime().TotalMilliseconds,
+                        ExecutionCount = FunctionExecutionCount,
+                        ExecutionTimeMS = FunctionExecutionTimeMS,
+                        IsAlwaysReady = IsAlwaysReady
+                    };
+
+                    FunctionExecutionTimeMS = FunctionExecutionCount = 0;
+                }
+
+                await PublishMetricsAsync(metrics);
+            }
+            finally
+            {
+                _intervalStopwatch = ValueStopwatch.StartNew();
+            }
+        }
+
+        private async void OnFunctionMetricsPublishTimer(object state)
+        {
+            await OnPublishMetrics();
+        }
+
+        private async Task PublishMetricsAsync(Metrics metrics)
+        {
+            string fileName = string.Empty;
+
+            try
+            {
+                bool metricsPublishEnabled = !string.IsNullOrEmpty(MetricsFilePath);
+                if (metricsPublishEnabled && !PrepareDirectoryForFile())
+                {
+                    return;
+                }
+
+                string metricsContent = JsonConvert.SerializeObject(metrics);
+                _logger.PublishingMetrics(metricsContent);
+
+                if (metricsPublishEnabled)
+                {
+                    fileName = $"{Guid.NewGuid().ToString().ToLower()}.json";
+                    string filePath = Path.Combine(MetricsFilePath, fileName);
+
+                    using (var streamWriter = _fileSystem.File.CreateText(filePath))
+                    {
+                        await streamWriter.WriteAsync(metricsContent);
+                    }
+                }
+            }
+            catch (Exception ex) when (!ex.IsFatal())
+            {
+                // TODO: consider using a retry strategy here
+                _logger.LogError(ex, $"Error writing metrics file '{fileName}'.");
+            }
+        }
+
+        private bool PrepareDirectoryForFile()
+        {
+            if (string.IsNullOrEmpty(MetricsFilePath))
+            {
+                return false;
+            }
+
+            // ensure the directory exists
+            _fileSystem.Directory.CreateDirectory(MetricsFilePath);
+
+            var metricsDirectoryInfo = _fileSystem.DirectoryInfo.FromDirectoryName(MetricsFilePath);
+            var files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // ensure we're under the max file count
+            if (files.Count < _optionsMonitor.CurrentValue.MaxFileCount)
+            {
+                return true;
+            }
+
+            // we're at or over limit
+            // delete enough files that we have space to write a new one
+            int numToDelete = files.Count - _optionsMonitor.CurrentValue.MaxFileCount + 1;
+            var filesToDelete = files.Take(numToDelete).ToArray();
```

Comment:
did we also consider the reader deleting the file once it has processed it?

Reply:
Yes, the reader will be deleting files when they're "consumed". This code is about making space to write a single new file. Do you see an issue here?

Reply:
so both writer and reader are deleting files?
its the same set of files ?

Reply:
Yes same set of files. Note that cleanup on this producer side only kicks in by default if there's 180 unprocessed files (15 minutes worth of files). That'd happen only if the consumer was stalled for a long period of time. When it comes back online you're right that the possibility a few conflicts might exist but they'd resolve quickly. E.g. consumer might list the directory and see 150 files to process, and while processing them it might find one or two were deleted. That's fine I think. The bulk of the files would be processed and we'd be back to steady state.

Reply:
got it. thanks.